src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,53 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
+    private Element<T> topOfStack;
+    private int size;
+
+    private static class Element<T> {
+        private Element<T> prev;
+        private T value;
+
+        public Element(Element<T> prev, T value) {
+            this.prev = prev;
```

Comment:
The constructor of the nested Element class is public. It should be private to ensure encapsulation as it is used only within the MateStack class.